### ISSUE TITLE
Custom trainable function in tf.scan within keras model

### ISSUE THREAD

**Issue by TC27GF** [tensorflow/tensorflow#52360]
I need to employ [tf.scan](https://www.tensorflow.org/api_docs/python/tf/scan) with a custom function (**fn**) in my Keras-based constructed model. The function contains a set of layers with some trainable weights; however, tf outputs some warnings which can not be ignored simply and need corrections. Please find below a very simplified code sample to reproduce the warnings.

------------------------

### System information

-   **Have I written custom code (as opposed to using a stock example script
    provided in TensorFlow)**: yes, a sample code is provided in the next.
-   **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 20.04
-   **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue
    happens on a mobile device**:
-   **TensorFlow installed from (source or binary)**: binary
-   **TensorFlow version (use command below)**: 2.6
-   **Python version**: Python 3.8.10
-   **Bazel version (if compiling from source)**:
-   **GCC/Compiler version (if compiling from source)**:
-   **CUDA/cuDNN version**: 11.1
-   **GPU model and memory**: 2080 ti
-   **Exact command to reproduce**: a sample code is provided in the next.

------------------------

### Sample code:

```

import tensorflow as tf

class fn(tf.keras.Model):
# class fn(tf.keras.layers.Layer):
  def __init__(self, units:int=512):
    super(fn, self).__init__()
    self.units = units
    self.dense1 = tf.keras.layers.Dense(self.units, name="dense1")
    self.dense2 = tf.keras.layers.Dense(self.units, name="dense2")
  def call(self, a, x):
    out = self.dense1(a) + self.dense2(x)
    return out
  


class solution:
  
  def __init__(self, a, b, units):
    self.a = a
    self.b = b
    self.units = units
    self.dense0 = tf.keras.layers.Dense(self.units, name="dense1")
    self.fn = fn(units=self.units)  
    self.make()
  
  def make(self):
    inputs = tf.keras.layers.Input(shape=[self.a, self.b],name="inputs")
    outputs = self.dense0(inputs)
    outputs = tf.scan(self.fn, outputs)
    # outputs = tf.scan(lambda a, x: self.fn(a,x), outputs)
    self.model = tf.keras.Model(inputs=inputs, outputs=outputs) 

model = solution(a=301, b=256, units=256).model

```

------------------------

### Model summary:
```
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
inputs (InputLayer)          [(None, 301, 256)]        0         
_________________________________________________________________
dense1 (Dense)               (None, 301, 256)          65792     
_________________________________________________________________
tf.scan_1 (TFOpLambda)       (None, 301, 256)          0         
=================================================================
Total params: 65,792
Trainable params: 65,792
Non-trainable params: 0
_________________________________________________________________
```

------------------------

### Warnings:

```
WARNING:tensorflow:
The following Variables were used a Lambda layer's call (tf.scan_1), but
are not present in its tracked objects:
  <tf.Variable 'tf.compat.v1.scan_1/scan/while/fn_1/dense1/kernel:0' shape=(256, 256) dtype=float32>
  <tf.Variable 'tf.compat.v1.scan_1/scan/while/fn_1/dense1/bias:0' shape=(256,) dtype=float32>
  <tf.Variable 'tf.compat.v1.scan_1/scan/while/fn_1/dense2/kernel:0' shape=(256, 256) dtype=float32>
  <tf.Variable 'tf.compat.v1.scan_1/scan/while/fn_1/dense2/bias:0' shape=(256,) dtype=float32>
It is possible that this is intended behavior, but it is more likely
an omission. This is a strong indication that this layer should be
formulated as a subclassed Layer rather than a Lambda layer.

```

------------------------

### Other links:

I also followed some other links, but not helpful: [link1](https://stackoverflow.com/questions/66873192/subclassing-a-keras-layer-the-following-variables-were-used-a-lambda-layers-ca) [link2](https://programming.vip/docs/day-6-tensorflow2-model-subclassing-api.html) [link3](https://stackoverflow.com/questions/67724151/keras-layers-multiheadattention-gives-warning-the-following-variables-were-used)

------------------------

### Stackoverflow link: 
(unfortunately, no feedback yet)

[Link](https://stackoverflow.com/questions/69546769/custome-trainable-function-in-tf-scan-variables-used-a-lambda-layers-call-are)

------------------------

**Deeply appreciate any help.**

**Comment by mohantym** [comment]
Hi @sanatmpa1 ! Could you look at this issue , it is replicating [2.5](https://colab.research.google.com/gist/mohantym/ef977648dc3a5ded6b554106c06d14b5/github_52360.ipynb#scrollTo=ar2v-yQ2336D) ,[2.6 ](https://colab.research.google.com/gist/mohantym/6836071ceedae8a5243880d70d2314d6/github_52360.ipynb#scrollTo=ar2v-yQ2336D)and [2.7](https://colab.research.google.com/gist/mohantym/04de6800c6da092a237083d50bf8588f/github_52360.ipynb#scrollTo=IE0pODYYm8eh)

**Comment by sachinprasadhs** [comment]
The warning indicates that the Variables mentioned in the warning are not part of calculating gradient, you can ignore the warning if you want it that way or you need to define a layer based on the base class instead of Lambda.

**Comment by TC27GF** [comment]
@sachinprasadhs thanks for the reply; unfortunately yes, the main problem was the gradient calculation for trainable weights that could not be ignored. Anyway, I couldn't find a good solution on the basis of above mentioned code, however, it was already solved using the following structure that might be helpful for others as well; the below structure can be employed for any complex trainable **fn** structure:

```
class Custom_scan_layer(tf.keras.layers.Layer):

  def __init__(self, params):
    super(Custom_scan_layer, self).__init__()
    self.params = params
    #======================================================================
    # define all trainable variables or layers which will be 
    # called/employed within the fn .
    #======================================================================

  def fn(self, a, x):
    #======================================================================
    # create your desired function using pre-defined variables or layers
    #======================================================================
    out = a + x 
    return out
  
  def call(self, inputs):
    out = tf.scan(self.fn, inputs)
    return out
```

**Comment by sachinprasadhs** [comment]
Good to hear that you have solved the issue, could you move this issue to closed. Thanks!

**Comment by SergioG-M** [comment]
I am getting warnings "WARNING:tensorflow:The following Variables were used in a Lambda layer's call" even if I set the env var. Using tf.get_logger().setLevel('ERROR') works though. Anyone else with the same issue?
